unicodetools/src/main/java/org/unicode/text/UCD/VersionedSymbolTable.java

```diff
@@ -0,0 +1,426 @@
+                            : beforeEquals.substring(posNotEqual + 1) + "=" + afterEquals;
+            leftHandSide = beforeEquals.substring(0, posNotEqual);
+            interiorlyNegated = true;
+        }
+        if (interiorlyNegated) {
+            final var complement = getNonNegatedPropertyQuerySet(leftHandSide, propertyPredicate);
+            result.addAll(complement.complement().removeAllStrings());
+        } else {
+            result.addAll(getNonNegatedPropertyQuerySet(leftHandSide, propertyPredicate));
+        }
+        return true;
+    }
+
+    private UnicodeSet getNonNegatedPropertyQuerySet(
+            String leftHandSide, String propertyPredicate) {
+        final var unqualifiedLeftHandSide = new StringBuilder(leftHandSide);
+        final var queriedVersion = parseVersionQualifier(unqualifiedLeftHandSide);
+        final var deducedQueriedVersion = queriedVersion == null ? implicitVersion : queriedVersion;
+
+        final var queriedProperties = IndexUnicodeProperties.make(deducedQueriedVersion);
+
+        if (propertyPredicate.length() == 0) {
+            // Either unary-property-query, or binary-property-query with an empty property-value.
+            final var script = queriedProperties.getProperty(UcdProperty.Script);
+            final var generalCategory = queriedProperties.getProperty(UcdProperty.General_Category);
+            if (script.isValidValue(unqualifiedLeftHandSide.toString())) {
+                return script.getSet(unqualifiedLeftHandSide.toString());
+            }
+            if (generalCategory.isValidValue(unqualifiedLeftHandSide.toString())) {
+                return getGeneralCategorySet(queriedProperties, unqualifiedLeftHandSide.toString());
+            }
+            UnicodeProperty queriedProperty =
+                    queriedProperties.getProperty(unqualifiedLeftHandSide.toString());
+            if (queriedProperty == null && unversionedExtensions != null) {
+                queriedProperty =
+                        unversionedExtensions.getProperty(unqualifiedLeftHandSide.toString());
+            }
+            if (queriedProperty == null) {
+                throw new IllegalArgumentException(
+                        "Invalid unary-query-expression; could not find property "
+                                + unqualifiedLeftHandSide);
+            }
+            if (!queriedProperty.isType(UnicodeProperty.BINARY_MASK)) {
+                // TODO(egg): Remove when we can tell this is a unary query.
+                if (queriedProperty.isType(UnicodeProperty.STRING_OR_MISC_MASK)) {
+                    return queriedProperty.getSet("");
+                }
+                throw new IllegalArgumentException(
+                        "Invalid unary-query-expression for non-binary property "
+                                + queriedProperty.getName());
+            }
+            return queriedProperty.getSet(UcdPropertyValues.Binary.Yes);
+        } else {
```

Comment:
optional: long function; maybe break it into multiple smaller ones? e.g., for most of the above call a getSomethingSetForEmptyValue(queriedProperties, leftHandSide)?